src/main/java/com/minecolonies/colony/buildings/BuildingBuilder.java

```diff
@@ -96,23 +120,131 @@ public String getJobName()
+        if(this.neededResources.containsKey(res))
+        {
+            preAmount = this.neededResources.get(res);
+        }
+        this.neededResources.put(res, preAmount + amount);
+        this.markDirty();
+    }
+
+    /**
+     * Reduce a resource of the needed list.
+     * @param res the resource.
+     * @param amount the amount.
+     */
+    public void reduceNeededResource(Block res, int amount)
+    {
+        int preAmount = 0;
+        if(this.neededResources.containsKey(res))
+        {
+            preAmount = this.neededResources.get(res);
+        }
+
+        if(preAmount - amount <= 0)
+        {
+            this.neededResources.remove(res);
+        }
+        else
+        {
+            this.neededResources.put(res, preAmount - amount);
+        }
+        this.markDirty();
+    }
+
+    /**
+     * Resets the needed resources completely.
+     */
+    public void resetNeededResources()
+    {
+        neededResources = new HashMap<>();
+        this.markDirty();
+    }
+
+    @Override
+    public void readFromNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.readFromNBT(compound);
+        final NBTTagList neededResTagList = compound.getTagList(TAG_RESOURCE_LIST, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < neededResTagList.tagCount(); ++i)
+        {
+            final NBTTagCompound neededRes = neededResTagList.getCompoundTagAt(i);
+            final IBlockState state = NBTUtil.readBlockState(neededRes);
+            final int amount = neededRes.getInteger(TAG_AMOUNT);
+            neededResources.put(state.getBlock(), amount);
+        }
+    }
+
+    @Override
+    public void writeToNBT(@NotNull final NBTTagCompound compound)
+    {
+        super.writeToNBT(compound);
+        @NotNull final NBTTagList neededResTagList = new NBTTagList();
+        for (@NotNull final Map.Entry<Block, Integer> entry : neededResources.entrySet())
+        {
+            @NotNull final NBTTagCompound neededRes = new NBTTagCompound();
+            NBTUtil.writeBlockState(neededRes, entry.getKey().getDefaultState());
+            neededRes.setInteger(TAG_AMOUNT, entry.getValue());
+
+            neededResTagList.appendTag(neededRes);
+        }
+        compound.setTag(TAG_RESOURCE_LIST, neededResTagList);
+    }
+
+    /**
+     * Method to serialize data to send it to the view.
+     *
+     * @param buf the used ByteBuffer.
+     */
+    @Override
+    public void serializeToView(@NotNull ByteBuf buf)
+    {
+        super.serializeToView(buf);
+
+        buf.writeInt(neededResources.size());
+
+        for (@NotNull final Map.Entry<Block, Integer> entry : neededResources.entrySet())
+        {
+            ByteBufUtils.writeUTF8String(buf, entry.getKey().getLocalizedName());
+            buf.writeInt(entry.getValue());
+        }
+    }
+
+    /**
      * Provides a view of the builder building class.
      */
     public static class View extends AbstractBuildingWorker.View
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this class from 51 to the maximum allowed 50 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2972)